qutip_lattice/lattice.py

```diff
@@ -200,1108 +257,6196 @@ def cell_structures(val_s=None, val_t=None, val_u=None):
 
-                    c_radius = (self.positions_of_sites[
-                            j]-self.positions_of_sites[i])/2
+                else:
+                    # calculate period
+                    for h in range(1, nSites+1, 1):
+                        DownState_S = np.roll(DownState, -h)
+
+                        if (DownState_S == DownState).all():
+                            pn = h
+                            break
+
+                    no_of_flips = 0
+                    DownState_shifted = DownState
+
+                    for m in range(nSites):
+                        DownState_m = np.roll(DownState, -m)
+                        DownState_m_int = np.sum(DownState_m * bin2dez, axis=0)
+
+                        DLT = np.argwhere(intDownStates == DownState_m_int)
+                        ind_down = DLT[:, 0][0]
+
+                        if m > 0:
+                            DownState_shifted = np.roll(DownState_shifted, -1)
+
+                            if np.mod(np.count_nonzero(DownState) + 1, 2):
+                                no_of_flips = no_of_flips + DownState_shifted[
+                                    -1]
+                        else:
+                            no_of_flips = 0
+
+                        NewRI = k + RowIndexUs
+                        NewCI = ind_down
+                        NewEn = np.sqrt(
+                            pn) / nSites * np.power(PP, no_of_flips) * np.exp(
+                                -2 * np.pi * 1J / nSites * ikVector * m)
+
+                        if Is == 0:
+                            UssRowIs = np.array([NewRI])
+                            UssColIs = np.array([NewCI])
+                            UssEntries = np.array([NewEn])
+                            Is = 1
+                        elif Is > 0:
+                            DLT = np.argwhere(UssRowIs == NewRI)
+                            iArg1 = DLT[:, 0]
+                            DLT = np.argwhere(UssColIs == NewCI)
+                            iArg2 = DLT[:, 0]
+                            if np.size(np.intersect1d(iArg1, iArg2)):
+                                AtR = np.intersect1d(iArg1, iArg2)
+                                UssEntries[AtR] = UssEntries[AtR] + NewEn
+                            else:
+                                Is = Is + 1
+                                UssRowIs = np.append(
+                                    UssRowIs, np.array([NewRI]), axis=0)
+                                UssColIs = np.append(
+                                    UssColIs, np.array([NewCI]), axis=0)
+                                UssEntries = np.append(
+                                    UssEntries, np.array([NewEn]), axis=0)
+
+            RowIndexUs = RowIndexUs + cumulIndex[ikVector+1] - cumulIndex[
+                ikVector]
+
+        Usss = sparse.csr_matrix((UssEntries, (UssRowIs, UssColIs)),
+                                 shape=(sumL, np.power(2, nSites)))
```

Comment:
Is creating these operator elements per element the only options?
Can't they be build using tensor product so we can follow the equation used to built it from reading the code?